cirq/ops/parallel_gate_operation.py

```diff
@@ -0,0 +1,233 @@
+
+    def _is_parameterized_(self) -> bool:
+        return protocols.is_parameterized(self.gate)
+
+    def _resolve_parameters_(self, resolver):
+        resolved_gate = protocols.resolve_parameters(self.gate, resolver)
+        return self.with_gate(resolved_gate)
+
+    def _formatted_exponent(self, info: protocols.CircuitDiagramInfo,
+                            args: protocols.CircuitDiagramInfoArgs
+                            ) -> Optional[str]:
+        if info.exponent == 0:
+            return '0'
+
+        # 1 is not shown.
+        if info.exponent == 1:
+            return None
+
+        # Round -1.0 into -1.
+        if info.exponent == -1:
+            return '-1'
+
+        # If it's a float, show the desired precision.
+        if isinstance(info.exponent, float):
+            if args.precision is not None:
+                # funky behavior of fraction, cast to str in constructor helps.
+                approx_frac = Fraction(info.exponent).limit_denominator(16)
+                if approx_frac.denominator not in [2, 4, 5, 10]:
+                    if abs(float(approx_frac)
+                           - info.exponent) < 10 ** -args.precision:
+                        return '({})'.format(approx_frac)
+
+                return '{{:.{}}}'.format(args.precision).format(info.exponent)
+        return repr(info.exponent)
+
+    def _circuit_diagram_info_(self,
+                               args: protocols.CircuitDiagramInfoArgs
+                               ) -> protocols.CircuitDiagramInfo:
+        diagram_info = protocols.circuit_diagram_info(self.gate,
+                                                      args,
+                                                      NotImplemented)
+        if diagram_info == NotImplemented:
+            return diagram_info
+
+        # Extract the formatted exponent and place it in wire symbols so that
+        # every gate on the diagram has an exponent instead of the last one
+        symbol = diagram_info.wire_symbols[0]
+        exponent = self._formatted_exponent(diagram_info, args)
+        if exponent is not None:
+            symbol = "{0}^{1}".format(symbol, exponent)
+
+        # Include every gate on the diagram instead of one
+        wire_symbols = (symbol,) * len(self.qubits)
+        # Set exponent=1 so that another extra exponent doesn't get tacked on
+        return protocols.CircuitDiagramInfo(wire_symbols=wire_symbols,
+                                            exponent=1,
+                                            connected=False)
+
+    def _phase_by_(self, phase_turns: float,
+                   qubit_index: int) -> 'ParallelGateOperation':
+        phased_gate = protocols.phase_by(self._gate, phase_turns, qubit_index,
+                                         default=None)
+        if phased_gate is None:
+            return NotImplemented
+        return self.with_gate(phased_gate)
+
+    def _trace_distance_bound_(self) -> float:
+        return 1.
```

Comment:
Delete this method if the bound is no better than the default bound.

Reply:
Agreed.